prereise/gather/demanddata/bldg_electrification/puma_hp_estimator.py

```diff
@@ -0,0 +1,373 @@
+        based on RECS and CBECS surveys
+    """
+
+    puma_fuel = const.puma_fuel
+    puma_data = const.puma_data
+    natgas_price = pd.read_csv(
+        os.path.join(
+            os.path.dirname(__file__), "data", "state_ng_residential_price_2019.csv"
+        ),
+        index_col=0,
+    )
+    puma_hp_df = pd.DataFrame(
+        {
+            "state": puma_fuel["state"],
+            "frac_elec_htg": puma_fuel["hh_elec"] / puma_fuel["hh_total"],
+            "log_hdd": np.log(puma_data["hdd65_normals"]),
+            "natgas_price": puma_fuel.apply(
+                lambda x: natgas_price.loc[x.loc["state"], "$/kWh"], axis=1
+            ),
+            "hh_total": puma_fuel["hh_total"],
+        },
+        index=puma_data.index,
+    )
+    puma_hp_df["frac_hp_fitted"] = puma_hp_df.apply(
+        lambda x: (
+            x["frac_elec_htg"] * hp_fit_df["s_frac_elec_htg"]
+            + x["log_hdd"] * hp_fit_df["s_log_hdd"]
+            + x["natgas_price"] * hp_fit_df["s_natgas_price"]
+            + hp_fit_df["const"]
+        )
+        if x["frac_elec_htg"] > 0.27
+        else (
+            x["frac_elec_htg"] * hp_fit_df_low_elec["s_frac_elec_htg"]
+            + x["frac_elec_htg"] ** 2 * hp_fit_df_low_elec["s_frac_elec_htg_sq"]
+            + hp_fit_df_low_elec["const"]
+        ),
+        axis=1,
+    )  # 0.27 is an empirical threshold
+    puma_hp_df["frac_hp_fitted"].clip(lower=0, inplace=True)
+    puma_hp_df["hh_hp"] = puma_hp_df["frac_hp_fitted"] * puma_hp_df["hh_total"]
+
+    # scale hp penetration to align with AHS metropolitan area data
+    puma_data_metro["frac_hp_fitted"] = puma_hp_df["frac_hp_fitted"]
+    puma_data_metro["hh_total"] = puma_hp_df["hh_total"]
+    puma_data_metro["hh_hp_fitted"] = (
+        puma_data_metro["frac_hp_fitted"] * puma_data_metro["hh_total"]
+    )
+    puma_data_metro["hh_hp_target"] = (
+        puma_data_metro["hp_penetration"] * puma_data_metro["hh_total"]
+    )
+    puma_metro = puma_data_metro.groupby("metro_id")[
+        ["hh_hp_fitted", "hh_hp_target"]
+    ].sum()
+    scaler = puma_metro["hh_hp_target"] / puma_metro["hh_hp_fitted"]
+
+    puma_data_metro.sort_values(
+        "frac_in_zone", ascending=False, inplace=True
+    )  # if a puma is in two metropolitan area, then scale it by the metro area that is larger overlaying with the puma
+    puma_data_metro = puma_data_metro[~puma_data_metro.index.duplicated(keep="first")]
+    puma_data_metro = puma_data_metro.sort_index()
+
+    puma_data_metro["frac_hp_scaled"] = puma_data_metro.apply(
+        lambda x: x.loc["frac_hp_fitted"] * scaler[x.loc["metro_id"]], axis=1
+    )
+    puma_hp_df.loc[
+        puma_hp_df.index.isin(puma_data_metro.index), "frac_hp_fitted"
+    ] = puma_data_metro["frac_hp_scaled"]
+    puma_hp_df["hh_hp"] = puma_hp_df["frac_hp_fitted"] * puma_hp_df["hh_total"]
+
+    # scale hp penetration to align with RECS and CBECS survey
+    hp_target = {}
```

Comment:
I guess we are calculating `res` and `com` independently thus not sure whether we will need this dict.

Reply:
got it